File: bank-service/src/main/java/rs/edu/raf/mapper/CreditTypeMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.CreditTypeCreateDTO;
import rs.edu.raf.dto.CreditTypeDTO;
import rs.edu.raf.model.credit.CreditType;
import rs.edu.raf.repository.CurrencyRepository;

@Component
@AllArgsConstructor
public class CreditTypeMapper {
    private CurrencyRepository currencyRepository;
    private CurrencyMapper currencyMapper;

    public CreditType creditTypeCreateDTOtoCreditType(CreditTypeCreateDTO creditTypeCreateDTO) {
        CreditType creditType = new CreditType();

        creditType.setName(creditTypeCreateDTO.name());
        creditType.setPrepayment(creditTypeCreateDTO.prepayment());
        creditType.setMinLoanTerm(creditTypeCreateDTO.minLoanTerm());
        creditType.setMaxLoanTerm(creditTypeCreateDTO.maxLoanTerm());
        creditType.setMaxLoanAmount(creditTypeCreateDTO.maxLoanAmount());
        creditType.setNominalInterestRate(creditTypeCreateDTO.nominalInterestRate());
        creditType.setCurrency(currencyRepository.findById(creditTypeCreateDTO.currencyId()).orElseThrow());

        return creditType;
    }

    public CreditTypeDTO creditTypeToCreditTypeDTO(CreditType creditType) {
        return new CreditTypeDTO(creditType.getId(), creditType.getName(), creditType.getNominalInterestRate(),
                creditType.getMinLoanTerm(), creditType.getMaxLoanTerm(), creditType.getMaxLoanAmount(),
                creditType.getPrepayment(), currencyMapper.currencyToCurrencyDTO(creditType.getCurrency()));
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/CreditTypeMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.dto.CreditTypeDTO
import org.example.bankservicekotlin.dto.CreditTypeCreateDTO
import org.example.bankservicekotlin.model.credit.CreditType
import org.example.bankservicekotlin.repository.CurrencyRepository
import org.springframework.stereotype.Component

@Component
class CreditTypeMapper(
    private val currencyRepository: CurrencyRepository, private val currencyMapper: CurrencyMapper
) {
    fun creditTypeCreateDTOtoCreditType(creditTypeCreateDTO: CreditTypeCreateDTO): CreditType {
        val currency = currencyRepository.findById(creditTypeCreateDTO.currencyId).orElseThrow()
        return CreditType(
            null,
            creditTypeCreateDTO.name,
            creditTypeCreateDTO.nominalInterestRate,
            creditTypeCreateDTO.minLoanTerm,
            creditTypeCreateDTO.maxLoanTerm,
            creditTypeCreateDTO.maxLoanAmount,
            creditTypeCreateDTO.prepayment,
            currency
        )
    }

    fun creditTypeToCreditTypeDTO(creditType: CreditType): CreditTypeDTO {
        return CreditTypeDTO(
            creditType.id,
            creditType.name,
            creditType.nominalInterestRate,
            creditType.minLoanTerm,
            creditType.maxLoanTerm,
            creditType.maxLoanAmount,
            creditType.prepayment,
            currencyMapper.currencyToCurrencyDTO(creditType.currency)
        )
    }
}